### DISCUSSION TITLE
Setting up a reaction package

### DISCUSSION THREAD

**Discussion by dimi3saad** [IDAES/idaes-pse#403] [Q&A] [answered]
Hello! I am trying to set up a reaction package, but I'm having a few difficulties. I am utilizing the following page in the documentation as my reference: https://idaes-pse.readthedocs.io/en/stable/user_guide/components/property_package/general_reactions/generic_reactions.html

I prefer to use the config dictionary at the moment, but it's not quite working, even if I copy paste things exactly from the documentation and assign random numbers to the parameters (ex: constant_dh_rxn = -21E3). Thus, even though the dictionary is correctly expressed and initialized, there seems to be a syntax error when I use: m.fs.reaction_properties = GenericReactionParameterBlock(default={ "property_package": m.fs.properties, config_dict }). The error relates to config_dict, and upon hovering over the red underlining of config_dict, I read: Dictionary entries must contain key/value pairs. Pylance. Could you please help me debug this issue?

In addition, I am not exactly sure how to express values for the variables in config_dict. I understand what the heat of reaction (constant_dh_rxn) would require, as it is a simple number, but what about rate_form (power_law_rate) for instance? I have not found an example to give me a baseline reference to what they could be filled in with. The documentation for these parameters is not too clear when it comes to assigning the corresponding values. I did understand that I would need to define a function for these parameters, but I would like an example on how to set up a function that would be accepted by the solvers. Could I get some sort of example for such values/parameters or functions?

Thank you in advance. I hope I was clear. Please let me know if you need more information!

**Comment by andrewlee94** [comment, marked as answer]
@dimi3saad I believe your problem is that you need to "unpack" the config dict (note the `**`):

`m.fs.reaction_properties = GenericReactionParameterBlock(default={ "property_package": m.fs.properties, **config_dict })`

From you questions about providing parameters for the config dict, I think you will need to be a bit more specific about what you are asking, as not all methods require parameters and there are different parameters for different forms. For power law type functions, there are no requires parameters, although there is a set of optional parameters. A power law form is:

`r = k*prod(x[i]**o[i] for i in component_list)`

For an elementary reaction, `o` can be calculated from the reaction stoichiometry, and thus is not a required parameter. However, for non-elementary reactions, you can supply an additional set of parameters for `o` by providing a dict of values for `reaction_order`.

**Reply by dimi3saad** [reply]
Hello Andrew,

Thank you for your reply! The unpacking of the config dict indeed worked. My problem now is also in that same line of code, wherein GenericReactionParameterBlock is not defined it seems. It is underlined with a yellow line, and hovering over it gives me that "it is not defined". I tried to address this by calling for the package using:
from idaes.generic_models.properties.core.generic import generic_reaction

I thought this would work, but it seems that the issue remains the same, and it is undefined, even though the import command shows no error. Could you please help me address this issue as well?

Concerning the parameters, I am still a bit unsure about how to address the parameters. You mentioned that some parameters are optional? For the rate_form, could I for example set power_law_rate = arrhenius * prod(x[i]**o[i] for i in component_list), with arrhenius being the input in "rate_constant": arrhenius? I have not tried this yet, but my concern is in how the dict will receive x[i] and o[i]. Should I input anything else? Finally, I understand what "concentration_form": ConcentrationForm.molarity refers to, but I am not sure the code acknowledges ConcentrationForm. It seems like I did not initialize it earlier. How do I do that in that case?

Apologies for the many questions, and thank you in advance! Please let me know if you need any further information.

**Reply by andrewlee94** [reply]
The import statement for generic reactions and ConcentrationForm is:

```
from idaes.generic_models.properties.core.generic.generic_reaction import GenericReactionParameterBlock, ConcentrationForm
```

For the reaction parameters, the best place to start is the example which can be found here:

https://github.com/IDAES/idaes-pse/blob/main/idaes/generic_models/properties/core/examples/reactions/reaction_example.py

First, `ConcentrationForm` can be imported from as shown above (it is an `Enum` which is defined in the generic reactions module).

Next, the `rate_form` is chosen by selecting from the pre-defined libraries (line 123 in example, lines 34&35 for associated import), or by providing a custom expression (I won't go into custom expressions yet).

Defining stoichiometry is shown in lines 118-120 in the example. You need to provide a dict of stoichiometric coefficients with the form `{(phase, component): coefficient}`. Any phase-component pair not defined is assumed to have a coefficient of 0. Defining stoichiometry is mandatory.

Additionally, you can chose to provide a separate dict of reaction orders if necessary. This is optional, and if it is not defined we assume an elementary reaction where order is equal to stoichiometry. Reaction order can be defined by adding a `parameter_data` entry to each reaction which contains a `reaction_order` entry which takes the form a dict like the stoichiometry. E.g.:

```
"rate_reactions": {
    "r1": {"stoichiometry": {("p1", "c1"): -1,
                                         ("p1", "c2"): 2},
    "rate_form": power_law_rate,
    "concentration_form": ConcentrationForm.moleFraction,
    "parameter_data": {
        "reaction_order": {("p1", "c1"): 1,
                                     ("p1", "c2"): 2,
                                     ("p2", "c1"): 3,
                                     ("p2", "c2"): 4}}}}
```

Updating and improving the documentation for generic reactions (and generic properties) is on our list of things to do by the end of this year.

**Reply by dimi3saad** [reply]
Hello Andrew,

Apologies for not getting back to you earlier, but I just wanted to confirm that your tips and answers did help me solve the problem, and now I have an operational reaction package (I used the stoichiometric reactor). Thank you so much!

**Comment by andrewlee94** [comment]
@dimi3saad Glad to hear that you got your example working.
